### ISSUE TITLE
[Bug]: The bootstrap package should be built along with the apk in F-Droid build

### ISSUE THREAD

**Issue by linsui** [termux/termux-app#2892]
### Problem description

Currently the bootstraps packages are downloaded at build time. Since F-Droid doesn't allow binaries, they should be built from source along with the apk. I thought we can use a srclib for the [termux-packages](https://github.com/termux/termux-packages). The problem is that we can't autoupdate it. A better way is include it as a submodule. Could you please help? Thanks!

### Steps to reproduce the behavior.

None

### What is the expected behavior?

The bootstraps packages are built along with the apk in F-Droid build.

### System information

* Termux application version:
* Android OS version:
* Device model:

**Comment by sylirre** [comment]
Packages are not part of the application. They are delivered from https://packages.termux.dev/. Bootstrap Archive is just a set of preinstalled packages from there and under normal conditions shouldn't contain binaries built by third parties (whether it is F-Droid or not).

Another issue is that we use specialized build environment. That's not a plain Android NDK setup for building native code. That's Ubuntu Docker image with extra tools installed, scripts for generating a patched standalone toolchain and building packages.

**Comment by linsui** [comment]
> Packages are not part of the application. They are delivered from https://packages.termux.dev/. Bootstrap Archive is just a set of preinstalled packages from there and under normal conditions shouldn't contain binaries built by third parties (whether it is F-Droid or not).

Yes, but it's shipped with the apk, right? Another way is to download the bootstraped package on the first run.

> Another issue is that we use specialized build environment. That's not a plain Android NDK setup for building native code. That's Ubuntu Docker image with extra tools installed, scripts for generating a patched standalone toolchain and building packages.

Does this require building the NDK? How difficult is it to setup such an environment? F-Droid uses Debian as the build server.

> Also, the build environment needs Internet access for downloading source code from the Internet. Not sure how this is relevant, but if F-Droid builds software on offline machine, that's won't work with Termux packages.

F-Droid doesn't build the apk on an offline machine, else the bootstraps packages can't be downloaded.

**Comment by Grimler91** [comment]
Our bootstraps are built by github actions in a fully transparent process: https://github.com/termux/termux-packages/blob/master/scripts/build-bootstraps.sh

Which is run weekly: https://github.com/termux/termux-packages/blob/master/.github/workflows/bootstrap_archives.yml

Has there been discussions in the fdroid community that termux should move away from downloading these bootstrap archives?

> Another way is to download the bootstraped package on the first run.

We used to do that, but this was problematic for users with bad/intermittent network connection

> How difficult is it to setup such an environment? F-Droid uses Debian as the build server.

The build environment would have to run an ubuntu container to build the bootstrap packages. App build time would increase by maybe half an hour (if all bootstrap packages should be built from source, rather than being downloaded from termux's repos). Building bootstraps could then also fail if for example a package's homepage is temporarily down, not sure if fdroid maintainers would be happy to investigate and solve such issues

**Comment by linsui** [comment]
> Has there been discussions in the fdroid community that termux should move away from downloading these bootstrap archives?

It's reported long time ago: https://gitlab.com/fdroid/fdroiddata/-/issues/2614. We do have some [exceptions](https://f-droid.org/en/docs/Inclusion_Policy/), but if it's not difficult to build the binary we should not add another one.

> The build environment would have to run an ubuntu container to build the bootstrap packages. App build time would increase by maybe half an hour (if all bootstrap packages should be built from source, rather than being downloaded from termux's repos). Building bootstraps could then also fail if for example a package's homepage is temporarily down, not sure if fdroid maintainers would be happy to investigate and solve such issues

Why does it have to be built in a Ubuntu container? We are building apps in Debian VM so it may be difficult to run a nest VM. Half an hour is not a problem. We have many less popular apps take hours to build.

**Comment by sylirre** [comment]
> Does this require building the NDK?

No, the NDK is standard. It is getting modified in build process.

> How difficult is it to setup such an environment? F-Droid uses Debian as the build server.

The main requirements are:
* Unrestricted Internet access.
* Ability to run *arbitrary scripts* as build entry point.
* Full root access to the system in order to setup and run Docker, or at least be able to install packages and create directories on rootfs.

Basically standard CI/CD environment. But F-Droid build process is *uncontrollable* and is designed to build Android applications only.

**Comment by linsui** [comment]
Let me tag some other F-Droid contributors: @IzzySoft @licaon-kter

**Comment by Grimler91** [comment]
> Why does it have to be built in a Ubuntu container? We are building apps in Debian VM so it may be difficult to run a nest VM

We use an ubuntu container to build packages, building in debian would require testing and probably some changes (like some host build dependencies/tools having different package names)

**Comment by Grimler91** [comment]
I could spend some time and do those tests and adjustments, but as @sylirre said fdroid anyways have to run arbritray scripts and commands to be able to build the packages using termux's build system

**Comment by sylirre** [comment]
> fdroid anyways have to run arbritray scripts

Technically solvable by writing `Android.mk` script usually used to build native code for APK. That will require some work and probably extra patches for packages that going to be added into bootstrap zip file. It may not use `/data/data/com.termux/files/usr` as installation prefix (e.g. `make install DESTDIR=/path/to/custom/dir` or with a patch).

But remaining questions are:
* It is possible to choose a version of NDK? Or can we use `curl` to download a specific version?
* Is it possible to run `sudo apt-get install <build_dependencies>` at least?
* Is F-Droid ok with downloading source code for packages from the Internet?

Though the best variant for us is to be able running a Docker container, like we do now on GitHub Actions.

**Comment by agnostic-apollo** [comment]
> Our bootstraps are built by github actions in a fully transparent process: https://github.com/termux/termux-packages/blob/master/scripts/build-bootstraps.sh
> 
> Which is run weekly: https://github.com/termux/termux-packages/blob/master/.github/workflows/bootstrap_archives.yml

Mistake there. Github action generates the bootstrap with https://github.com/termux/termux-packages/blob/master/scripts/generate-bootstraps.sh and does not build it. Packages are built and published on our repo from github actions. The `generate-bootstrap.sh` just downloads already built packages, so is fast.

https://github.com/termux/termux-packages/wiki/For-maintainers#generating-bootstrap-archives 

The `build-bootstraps.sh` could be used for F-Droid, but there are some issues that need solving, will look into them in coming weeks. One major issue is that `dpkg` required by bootstrap also builds `dpkg-perl` subpackage, which depends on `clang` subpackage provided by `libllvm`, whose last build on github took 3hrs 48min, takes lot of time locally too. The `dpkg-perl` itself is not packaged in bootstrap so it does not need to be built, and so neither does `libllvm`. I added support locally to prevent such subpackages to be built if subpackage is not a dependency of parent package or sibling subpackages, but need to add support for when a subpackage may be needed by another package in bootstrap list.

https://github.com/termux/termux-packages/blob/master/packages/dpkg/dpkg-perl.subpackage.sh

https://github.com/termux/termux-packages/blob/master/packages/libllvm/clang.subpackage.sh

https://github.com/termux/termux-packages/actions/runs/2602567063

F-Droid is using gitlab actions for everything, so there would be monthly time limits too. Setting up build environment, downloading sources, building packages may take a lot of time for this to be viable. Although termux app releases are not done quite so often (because of me).

And considering that package source urls are removed by hosters quite often, we will need to check before each release if all urls are still valid, otherwise build will fail, also anyone trying to build old version of termux with f-droid may fail as well.

**Comment by IzzySoft** [comment]
@linsui what input can I provide? If I see this correctly, this is about packaging and build environments, which is out of my scope (remember, I'm not an Android dev).

**Comment by Grimler91** [comment]
> Mistake there. Github action generates the bootstrap with https://github.com/termux/termux-packages/blob/master/scripts/generate-bootstraps.sh and does not build it. Packages are built and published on our repo from github actions. The `generate-bootstrap.sh` just downloads already built packages, so is fast.

Yeah, my bad, I was thinking we used the same script but with an -i option for fast building, as for build-package.sh

**Comment by agnostic-apollo** [comment]
Ah. That could work too I guess. No worries.

**Comment by linsui** [comment]
The fdroid buildserver is basically a Debian server. We don't have an app built with Docker but has no problem with other requirements.

**Comment by linsui** [comment]
@IzzySoft Should we make another exception for Termux?

**Comment by linsui** [comment]
> No, the NDK is standard. It is getting modified in build process.

We can modify the NDK at build time.

> I could spend some time and do those tests and adjustments, but as @sylirre said fdroid anyways have to run arbritray scripts and commands to be able to build the packages using termux's build system

We can run anything on Debian as long as it's FOSS. We need to run sudo command seperately (in `sudo` block).

> Technically solvable by writing Android.mk script usually used to build native code for APK. That will require some work and probably extra patches for packages that going to be added into bootstrap zip file. It may not use /data/data/com.termux/files/usr as installation prefix (e.g. instead can do something like make install DESTDIR=/path/to/custom/dir or Makefile install rules can be patched).

We can just use a bash script.

> Is it possible to choose a version of NDK? Or can we use curl to download a specific version?

You can specify the NDK version. Downloading with curl is OK.

> Is it possible to run sudo apt-get install <build_dependencies> at least?

Of course.

> Is F-Droid ok with downloading source code for packages from the Internet?

Yes, as long as it's source code. It's better that downloading binaries from the internet, right?

> F-Droid is using gitlab actions for everything, so there would be monthly time limits too. Setting up build environment, downloading sources, building packages may take a lot of time for this to be viable. Although termux app releases are not done quite so often (because of me).

It's used for CI only. The real build is on a buildserver with Debian VMs.

> And considering that package source urls are removed by hosters quite often, we will need to check before each release if all urls are still valid, otherwise build will fail, also anyone trying to build old version of termux with f-droid may fail as well.

I guess this can be solved by using git instead. But it may takes more time.

**Comment by licaon-kter** [comment]
Let's see how to build it first I guess :)

**Comment by IzzySoft** [comment]
@linsui I have a hard time following the discussion. What kind of exception? I cannot tell if we can use an Ubuntu Docker image to build an app (or any other Docker images at that), if that's considered an exception (from what rule?) – you and Licaon are better able to tell. I vaguely remember we had a similar discussion in the past (something with a requirement towards specific Docker containers) and decided not to – but I don't remember the details as that's not my expertise. So maybe I rather unsubscribe from this thread, and have @licaon-kter adding his expertise? I'll do so now; if you think there's a question I *can* answer, feel free to ping me again so I get a notification.

**Comment by agnostic-apollo** [comment]
>It's used for CI only. The real build is on a buildserver with Debian VMs.

Ah, right. Thanks for that. I guess we can look into this and see how it goes, since building bootstrap on F-Droid servers would probably be ideal if it can actually work. The `build.gradle` would need updating as well to support file paths instead of url. Will also need to fix some `build-bootstrap.sh` issues, planned before next release.

>Let's see how to build it first I guess :)

If F-Droid will allow downloading and running our docker container we use to build packages, then simply `build-bootstrap.sh` could be used. Considering we will have `sudo` access, it should be technically possible, unless against policy. But docker container is currently `3.7GB` in size, so would that be an issue? Having to create a separate build system just for F-Droid may create its own specific issues too and increase maintenance work, so would be preferable to avoid that. And having same build systems would provide consistency with Github and other releases as well. If we can get this to work, possibly other F-Droid released apps with in-built binaries could use a similar setup.

https://hub.docker.com/r/termux/package-builder

https://github.com/termux/termux-packages/blob/master/scripts/Dockerfile

https://github.com/termux/termux-packages/blob/master/scripts/run-docker.sh

https://github.com/termux/termux-packages/blob/master/build-package.sh

https://github.com/termux/termux-packages/blob/master/scripts/build-bootstraps.sh
